### PR TITLE
KAFKA-8730: Add API to delete consumer offsets (KIP-496)

### PR DESCRIPTION
**Pull request by dajac**
This adds an administrative API to delete consumer offsets of a group as well as extends the mechanism to expire offsets of consumer groups.

It makes the group coordinator aware of the set of topics a consumer group (protocol type == 'consumer') is actively subscribed to, allowing offsets of topics which are not actively subscribed to by the group to be either expired or administratively deleted. The expiration rules remain the same.

For the other groups (non-consumer), the API allows to delete offsets when the group is empty and the expiration remains the same.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+public class GroupSubscribedToTopicException extends ApiException {
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: This is a public interface and wasn't mentioned in the KIP. Was there an offline discussion on adding this?

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: Right. There was not an offline discussion about it. I have forgotten to bring the point. What's the best way? Shall I discuss thing in the voting thread?

To give the context: When one tried to delete the offsets of an active consumer group, we need a way to indicate if the topic is still actively consumed by the group which means that it is not possible to delete the offsets.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: @hachikuji What do you think about adding this one?

Comment (gwenshap) [reply clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: I think this makes sense, and you can bring it up in the vote thread and ask if anyone objects.

Comment (gwenshap) [reply clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: Good catch @stanislavkozlovski :)

**Comment by gwenshap** [comment]
Is there a reason we are not using the protocol generation for the new protocol?

**Comment by dajac** [comment]
> Is there a reason we are not using the protocol generation for the new protocol?

it is used. See the json files which generates OffsetDeleteRequestData and OffsetDeleteResponseData which are encapsulated respectively in OffsetDeleteRequest and OffsetDeleteResponse. I have followed what has been done for other requests/responses: e.g. https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java

File: clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    public KafkaFuture<Void> partitionResult(final TopicPartition partition) {
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+
+        this.future.whenComplete(new BiConsumer<Map<TopicPartition, Errors>, Throwable>() {
+            @Override
+            public void accept(final Map<TopicPartition, Errors> topicPartitions, final Throwable throwable) {
+                if (throwable != null) {
+                    result.completeExceptionally(throwable);
+                } else if (!topicPartitions.containsKey(partition)) {
+                    result.completeExceptionally(new UnknownTopicOrPartitionException(
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java]: This handles the case of a topic that wasn't in the requested set of partitions? Seems like `IllegalArgumentException` might be more appropriate?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java [hunk 3]
```diff
@@ -120,6 +120,11 @@
         new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
         new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
 
+    public static Short deserializeHeader(ByteBuffer buffer) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java]: nit: maybe `deserializeVersion` so that the result is part of the name?

File: clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:19]: Seems reasonable to me.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 4]
```diff
@@ -312,7 +313,9 @@
             EligibleLeadersNotAvailableException::new),
     ELECTION_NOT_NEEDED(84, "Leader election not needed for topic partition", ElectionNotNeededException::new),
     NO_REASSIGNMENT_IN_PROGRESS(85, "No partition reassignment is in progress.",
-            NoReassignmentInProgressException::new);
+            NoReassignmentInProgressException::new),
+    GROUP_SUBSCRIBED_TO_TOPIC(86, "The consumer group is actively subscribed to the topic",
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Might be worth adding a little more detail about the context for this error.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 5]
```diff
@@ -3053,6 +3060,98 @@ void handleFailure(Throwable throwable) {
...
+                final Map<TopicPartition, Errors> partitions = new HashMap<>();
+                response.data.topics().forEach(topic -> {
+                    topic.partitions().forEach(partition -> {
+                        partitions.put(
+                            new TopicPartition(topic.name(), partition.partitionIndex()),
+                            Errors.forCode(partition.errorCode()));
+                    });
+                });
+
+                // If the error is an error at the group level, the future is failed with it
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Some other group level APIs have a top-level error code (e.g. OffsetFetchResponse). Do you think it's worth adding one?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 6]
```diff
@@ -3053,6 +3060,98 @@ void handleFailure(Throwable throwable) {
...
+                if (!partitions.isEmpty()) {
+                    final Errors error = partitions.values().iterator().next();
+                    if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS || error == Errors.COORDINATOR_NOT_AVAILABLE) {
+                        throw error.exception();
+                    } else if (error == Errors.INVALID_GROUP_ID || error == Errors.GROUP_ID_NOT_FOUND) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Another one is `GROUP_AUTHORIZATION_FAILED`.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        this.data = data;
+    }
+
+    public OffsetDeleteRequest(Struct struct, short version) {
+        super(ApiKeys.OFFSET_DELETE, version);
+        this.data = new OffsetDeleteRequestData(struct, version);
+    }
+
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Errors error) {
+        switch (version()) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java]: Hmm. I think we're just following the same pattern used elsewhere, but I think the version check is redundant. We already know it's a valid version because the request constructor verifies it.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.kafka.common.protocol.types.Struct;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Possible error codes:
+ * <p>
+ * COORDINATOR_LOAD_IN_PROGRESS (14)
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java]: nit: Can we link to the error code as we do in `OffsetFetchResponse`?

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public int throttleTimeMs() {
+        return data.throttleTimeMs();
+    }
+
+    @Override
+    public boolean shouldClientThrottle(short version) {
+        return version >= 1;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java]: Shouldn't this be 0?

File: clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java [hunk 10]
```diff
@@ -1517,6 +1526,152 @@ public void testDeleteConsumerGroups() throws Exception {
...
+                Collections.<String>emptySet(), nodes.get(0));
+
+        final String groupId = "group-0";
+        final TopicPartition tp1 = new TopicPartition("foo", 0);
+        final TopicPartition tp2 = new TopicPartition("bar", 0);
+
+        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(cluster)) {
+            env.kafkaClient().setNodeApiVersions(NodeApiVersions.create());
+
+            //Retriable FindCoordinatorResponse errors should be retried
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java]: Any chance we could break this test up? I think each of these comments could probably be a separate test case.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. I made a pass on the client logic. I will look at the broker implementation next.

File: clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java (hunk 2 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java]: You're right. I say `UnknownTopicOrPartitionException` here [1] so I thought it was the way to go in the Admin client. `IllegalArgumentException` is better.

[1] https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/ElectPreferredLeadersResult.java#L62

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java (hunk 3 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java]: I agree. I will change it.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 4 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Ack. I will make it better.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java (hunk 8 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java]: Sure thing.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java (hunk 9 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java]: Indeed. Good catch.

File: clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java (hunk 10 above)
Comment (dajac) [reply clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java]: Yes. I have should have done this before.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java (hunk 7 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java]: I have checked and your are right. I have removed it.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 5 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Yeah. It is worth adding one. It will avoid having to handle global errors manually in the admin client. Also, it feels cleaner.

**Comment by dajac** [comment]
@hachikuji I have addressed all your comments. I have refactored the response to include a top level error. It is cleaner like this. We need to bring the change in the KIP. I will do it if you are fine with the changes.

**Comment by hachikuji** [comment]
@dajac Sounds good to me. I'll try to review the rest today.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 11]
```diff
@@ -95,6 +96,26 @@ class GroupMetadataManager(brokerId: Int,
     "group-coordinator-metrics",
     "The avg time it took to load the partitions in the last 30sec"), new Avg())
 
+  val offsetCommitsSensor = metrics.sensor("OffsetCommits")
+
+  offsetCommitsSensor.add(new Meter(
+    metrics.metricName("offset-commits-rate",
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: maybe `offset-commit-rate`?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala [hunk 12]
```diff
@@ -423,6 +434,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
...
+            members.map { case (_, member) =>
+              val buffer = ByteBuffer.wrap(member.metadata(protocol.get))
+              ConsumerProtocol.deserializeVersion(buffer)
+              ConsumerProtocol.deserializeSubscriptionV0(buffer).topics().asScala.toSet
+            }.reduceLeft(_ ++ _)
+          )
+        } catch {
+          case e: SchemaException => {
+            warn(s"Failed to parse Consumer Protocol ${ConsumerProtocol.PROTOCOL_TYPE}:${protocol.get} " +
+              s"of group $groupId. Consumer Group is not aware of the subscribed topics.", e)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: Consumer group _coordinator_ is not aware?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 13]
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def handleOffsetDeleteRequest(request: RequestChannel.Request): Unit = {
+    val offsetDeleteRequest = request.body[OffsetDeleteRequest]
+    val groupId = offsetDeleteRequest.data.groupId()
+
+    if (authorize(request.session, Read, Resource(Group, groupId, LITERAL))) {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I guess we didn't specify this in the KIP, but it seems reasonable to use the same permission as the OffsetCommit API. On the other hand, the DeleteGroups API requires Delete permission. Since they are both administrative use cases, do you think it makes sense for them to be consistent?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 14]
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      val topicPartitions = offsetDeleteRequest.data.topics().asScala.flatMap { topic =>
+        topic.partitions().asScala.map { partition =>
+          new TopicPartition(topic.name(), partition.partitionIndex())
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: The OffsetCommit and OffsetFetch APIs check permissions on each topic included in the request. Should we do the same here? I think we're basically trying to avoid letting this API leak information about the topics in the cluster.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala [hunk 15]
```diff
@@ -423,6 +434,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
...
+   */
+  private[group] def computeSubscribedTopics(): Option[Set[String]] = {
+    protocolType match {
+      case Some(ConsumerProtocol.PROTOCOL_TYPE) if !members.isEmpty && protocol.isDefined =>
+        try {
+          Some(
+            members.map { case (_, member) =>
+              val buffer = ByteBuffer.wrap(member.metadata(protocol.get))
+              ConsumerProtocol.deserializeVersion(buffer)
+              ConsumerProtocol.deserializeSubscriptionV0(buffer).topics().asScala.toSet
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: It may be worth a comment explaining why we deserialize as version 0.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 16]
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      }.toSeq
+
+      val (groupError, topicErrors) = groupCoordinator.handleDeleteOffsets(groupId, topicPartitions)
+
+      sendResponseMaybeThrottle(request, requestThrottleMs => {
+        if (groupError != Errors.NONE)
+          offsetDeleteRequest.getErrorResponse(requestThrottleMs, groupError)
+        else {
+          val topics = new OffsetDeleteResponseData.OffsetDeleteResponseTopicCollection
+          topicErrors.groupBy(_._1.topic()).map { case (topic, topicPartitions) =>
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: it's conventional in scala to leave off parenthesis for simple getters like `topic()`. There are a few similar cases below

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala [hunk 17]
```diff
@@ -204,6 +209,10 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   private var receivedTransactionalOffsetCommits = false
   private var receivedConsumerOffsetCommits = false
 
+  // When protocolType == `consumer`, a set of subscribed topics is maintained. The set is
+  // computed when a new generation is created or when the group is restored from the log.
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:213]: So this means that offset deletion does not take into account the subscriptions of an active rebalance. That's probably fine I guess since it is simpler and we are not modifying the commit state for any partitions which are actively being fetched.

A couple questions for completeness:

1. For the sake of argument, suppose we loosened the validation so that we allow explicit deletions (but not expirations) without respect to the group or subscription state. The behavior of the consumer would be a little unpredictable. It might just write over the deletion. It might also force an offset reset if an OffsetFetch followed the deletion. Because the behavior is undefined, we do not allow it, which seems fair. However, the downside is that it makes the API a little harder to use, so just wanted to check whether you think it is a worthwhile restriction.

2. If we know the subscription, then it wouldn't be much trouble to delete the offsets of all partitions that are not currently subscribed to. This could be supported in the API by using a null in the OffsetDelete request. I think it could be useful, but I'm inclined to leave it for future work. What do you think?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 18]
```diff
@@ -519,6 +542,62 @@ class GroupCoordinator(val brokerId: Int,
     groupErrors
   }
 
+  def handleDeleteOffsets(groupId: String, partitions: Seq[TopicPartition]): (Errors, Map[TopicPartition, Errors]) = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:545]: Leaving this here for lack of a better alternative. One thing that I didn't cover in the KIP is changes to the consumer group command to use this API. I think we can extend the KIP and do this in a separate PR. What do you think?

**Review by hachikuji — COMMENTED** [review]
Thanks, the broker logic looks pretty good. I left a few comments/questions.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 13 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yeah. I thought it makes more sense to allow an application to delete an offset if it can commit it. On the other hand, we could argue that deleting is more an exceptional behaviour so it could make sense to have the possibility to restrict it. I think the later is more reasonable. I will update it.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 14 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: We should do the same. I naively thought that all members of a group would all have the permissions to see the topic used in the group. It seems it is not true. I will add the check as you suggest.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala (hunk 17 above)
Comment (dajac) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:213]: Yes and No. I have distinguished two cases when I designed this:

1) When the group is empty and the initial rebalance takes place, subscriptions will be ignored until the rebalance is completed and a new generation is initialised. I thought it would be inappropriate to compute the set of subscribed topics before having the full view of the consumer group. 

2) When the group is not empty and a rebalance takes place, the current set of subscribed topics will remain until all rebalance is completed and a new generation is initialised. I think this is a good think to be conservative here until the group has a full view of all the members.

Another aspect is that the protocol is actually selected at this point as well. I think it is better to stay consistent and parse the consumer protocol once the protocol of the group has been selected. It is not a hard requirement though as we could parse any versions but it feels more consistent in my opinion.

To answer your questions:

1) I think allowing explicit deletions without respect to the group or subscription state would be really risky as the behaviour of the consumer would be unpredictable as you pointed out. I would say that it does not make much sense anyway to delete offsets of a topic while a running consumer group consumes this topic. It is good to have the restriction in place in my opinion.

2) Yeah. It would be possible to delete the offsets of all partitions *of the topics* that are not currently subscribed to. I wonder it it really useful though as they will be eventually cleaned up now. I don't see the benefit of having the possibility to basically force the immediate clean up. I would leave it for future work and re-iterate based on the feedback we get from the users.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 18 above)
Comment (dajac) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:545]: It makes sense. Let's do it in a separate PR.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 14 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: In the same time, I have added a check which verifies if the group knows the partition. `UNKNOWN_TOPIC_OR_PARTITION` is returned if the partition is unknown.

**Comment by dajac** [comment]
@hachikuji I think I have addressed all your questions and comments. Thanks for the review! I will rebase the PR to resolve conflicts once you are done with the review.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 18 above)
Comment (dajac) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:545]: I have opened a JIRA for it: https://issues.apache.org/jira/browse/KAFKA-8901

**Comment by dajac** [comment]
retest this please

**Comment by dajac** [comment]
retest this please

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala (hunk 17 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:213]: Thanks for the response. Makes sense and I agree.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+    public OffsetDeleteRequest(Struct struct, short version) {
+        super(ApiKeys.OFFSET_DELETE, version);
+        this.data = new OffsetDeleteRequestData(struct, version);
+    }
+
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Errors error) {
+        OffsetDeleteResponseTopicCollection topics = new OffsetDeleteResponseTopicCollection();
+
+        for (OffsetDeleteRequestTopic topic : data.topics()) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java]: In this case, would it be reasonable to set only the top level error? If the client sees a top-level error code which is not NONE, then there is no need to process partition-level errors.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 20]
```diff
@@ -95,6 +96,26 @@ class GroupMetadataManager(brokerId: Int,
...
     "The avg time it took to load the partitions in the last 30sec"), new Avg())
 
+  val offsetCommitsSensor = metrics.sensor("OffsetCommits")
+
+  offsetCommitsSensor.add(new Meter(
+    metrics.metricName("offset-commit-rate",
+      "group-coordinator-metrics",
+      "The rate of committed offsets"),
+    metrics.metricName("offset-commit-count",
+    "group-coordinator-metrics",
```
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: alignment is a little inconsistent here

**Review by hachikuji — COMMENTED** [review]
Just one more comment. This is looking good.

File: clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java (hunk 19 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java]: Totally. I have missed this one.

**Comment by dajac** [comment]
@hachikuji Thank you for the review. I have addressed the remaining points. The builds keep being killed after 240 minutes. Is it a hard limit set in Jenkins?

**Comment by hachikuji** [comment]
@dajac I added another 30 minutes to the build time. Let's see if that's enough.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the PR!
